File: data_structures/linked_list/go/linked_list.go
package datastructures

import (
	"errors"
)

// ============================================================
// ListNode — simple linked list node used by problem solutions
// ============================================================

// ListNode is a simple linked list node used by problem solutions.
type ListNode struct {
	Val  int
	Next *ListNode
}

// Arr2Node converts a slice of ints into a linked list and returns the head.
func Arr2Node(input []int) *ListNode {
	if len(input) == 0 {
		return nil
	}

	head := &ListNode{
		Val:  input[0],
		Next: nil,
	}
	curr := head

	for i := 1; i < len(input); i++ {
		curr.Next = &ListNode{
			Val:  input[i],
			Next: nil,
		}

		curr = curr.Next
	}

	return head
}

var (
	ErrEmptyList      = errors.New("list is empty")
	ErrIndexOutOfRange = errors.New("index out of range")
)

// ============================================================
// Singly Linked List
// ============================================================

type SinglyNode struct {
	Val  int
	Next *SinglyNode
}

type SinglyLinkedList struct {
	Head *SinglyNode
	size int
}

func NewSinglyLinkedList() *SinglyLinkedList {
	return &SinglyLinkedList{}
}

func (l *SinglyLinkedList) InsertAtHead(val int) {
	node := &SinglyNode{Val: val, Next: l.Head}
	l.Head = node
	l.size++
}

func (l *SinglyLinkedList) InsertAtTail(val int) {
	node := &SinglyNode{Val: val}

	if l.Head == nil {
		l.Head = node
		l.size++
		return
	}

	curr := l.Head
	for curr.Next != nil {
		curr = curr.Next
	}
	curr.Next = node
	l.size++
}

func (l *SinglyLinkedList) InsertAt(index, val int) error {
	if index < 0 || index > l.size {
		return ErrIndexOutOfRange
	}

	if index == 0 {
		l.InsertAtHead(val)
		return nil
	}

	curr := l.Head
	for i := 0; i < index-1; i++ {
		curr = curr.Next
	}

	node := &SinglyNode{Val: val, Next: curr.Next}
	curr.Next = node
	l.size++
	return nil
}

func (l *SinglyLinkedList) DeleteAtHead() (int, error) {
	if l.Head == nil {
		return 0, ErrEmptyList
	}

	val := l.Head.Val
	l.Head = l.Head.Next
	l.size--
	return val, nil
}

func (l *SinglyLinkedList) DeleteAtTail() (int, error) {
	if l.Head == nil {
		return 0, ErrEmptyList
	}

	if l.Head.Next == nil {
		val := l.Head.Val
		l.Head = nil
		l.size--
		return val, nil
	}

	curr := l.Head
	for curr.Next.Next != nil {
		curr = curr.Next
	}

	val := curr.Next.Val
	curr.Next = nil
	l.size--
	return val, nil
}

func (l *SinglyLinkedList) DeleteAt(index int) (int, error) {
	if l.Head == nil {
		return 0, ErrEmptyList
	}

	if index < 0 || index >= l.size {
		return 0, ErrIndexOutOfRange
	}

	if index == 0 {
		return l.DeleteAtHead()
	}

	curr := l.Head
	for i := 0; i < index-1; i++ {
		curr = curr.Next
	}

	val := curr.Next.Val
	curr.Next = curr.Next.Next
	l.size--
	return val, nil
}

func (l *SinglyLinkedList) Search(val int) int {
	curr := l.Head
	for i := 0; curr != nil; i++ {
		if curr.Val == val {
			return i
		}
		curr = curr.Next
	}
	return -1
}

func (l *SinglyLinkedList) Reverse() {
	var prev *SinglyNode
	curr := l.Head

	for curr != nil {
		next := curr.Next
		curr.Next = prev
		prev = curr
		curr = next
	}

	l.Head = prev
}

func (l *SinglyLinkedList) Len() int {
	return l.size
}

func (l *SinglyLinkedList) ToSlice() []int {
	result := make([]int, 0, l.size)
	curr := l.Head
	for curr != nil {
		result = append(result, curr.Val)
		curr = curr.Next
	}
	return result
}

// ============================================================
// Doubly Linked List
// ============================================================

type DoublyNode struct {
	Val  int
	Prev *DoublyNode
	Next *DoublyNode
}

type DoublyLinkedList struct {
	Head *DoublyNode
	Tail *DoublyNode
	size int
}

func NewDoublyLinkedList() *DoublyLinkedList {
	return &DoublyLinkedList{}
}

func (l *DoublyLinkedList) InsertAtHead(val int) {
	node := &DoublyNode{Val: val, Next: l.Head}

	if l.Head != nil {
		l.Head.Prev = node
	} else {
		l.Tail = node
	}

	l.Head = node
	l.size++
}

func (l *DoublyLinkedList) InsertAtTail(val int) {
	node := &DoublyNode{Val: val, Prev: l.Tail}

	if l.Tail != nil {
		l.Tail.Next = node
	} else {
		l.Head = node
	}

	l.Tail = node
	l.size++
}

func (l *DoublyLinkedList) DeleteAtHead() (int, error) {
	if l.Head == nil {
		return 0, ErrEmptyList
	}

	val := l.Head.Val
	l.Head = l.Head.Next

	if l.Head != nil {
		l.Head.Prev = nil
	} else {
		l.Tail = nil
	}

	l.size--
	return val, nil
}

func (l *DoublyLinkedList) DeleteAtTail() (int, error) {
	if l.Tail == nil {
		return 0, ErrEmptyList
	}

	val := l.Tail.Val
	l.Tail = l.Tail.Prev

	if l.Tail != nil {
		l.Tail.Next = nil
	} else {
		l.Head = nil
	}

	l.size--
	return val, nil
}

func (l *DoublyLinkedList) Search(val int) int {
	curr := l.Head
	for i := 0; curr != nil; i++ {
		if curr.Val == val {
			return i
		}
		curr = curr.Next
	}
	return -1
}

func (l *DoublyLinkedList) Reverse() {
	curr := l.Head

	for curr != nil {
		curr.Prev, curr.Next = curr.Next, curr.Prev
		curr = curr.Prev
	}

	l.Head, l.Tail = l.Tail, l.Head
}

func (l *DoublyLinkedList) Len() int {
	return l.size
}

func (l *DoublyLinkedList) ToSlice() []int {
	result := make([]int, 0, l.size)
	curr := l.Head
	for curr != nil {
		result = append(result, curr.Val)
		curr = curr.Next
	}
	return result
}

File: data_structures/linked_list/python/linked_list.py
from __future__ import annotations


class ListNode:
    """Simple linked list node used by problem solutions."""
    def __init__(self, val: int = 0, next: ListNode | None = None):
        self.val = val
        self.next = next


def arr2node(values: list[int]) -> ListNode | None:
    if not values:
        return None
    head = ListNode(values[0])
    curr = head
    for v in values[1:]:
        curr.next = ListNode(v)
        curr = curr.next
    return head


# ── Singly Linked List ───────────────────────────────────────

class SinglyNode:
    def __init__(self, val: int = 0, next: SinglyNode | None = None):
        self.val = val
        self.next = next


class SinglyLinkedList:
    def __init__(self):
        self.head: SinglyNode | None = None
        self._size = 0

    def insert_at_head(self, val: int) -> None:
        self.head = SinglyNode(val, self.head)
        self._size += 1

    def insert_at_tail(self, val: int) -> None:
        node = SinglyNode(val)
        if self.head is None:
            self.head = node
            self._size += 1
            return
        curr = self.head
        while curr.next is not None:
            curr = curr.next
        curr.next = node
        self._size += 1

    def insert_at(self, index: int, val: int) -> None:
        if index < 0 or index > self._size:
            raise IndexError("index out of range")
        if index == 0:
            self.insert_at_head(val)
            return
        curr = self.head
        for _ in range(index - 1):
            curr = curr.next
        curr.next = SinglyNode(val, curr.next)
        self._size += 1

    def delete_at_head(self) -> int:
        if self.head is None:
            raise IndexError("list is empty")
        val = self.head.val
        self.head = self.head.next
        self._size -= 1
        return val

    def delete_at_tail(self) -> int:
        if self.head is None:
            raise IndexError("list is empty")
        if self.head.next is None:
            val = self.head.val
            self.head = None
            self._size -= 1
            return val
        curr = self.head
        while curr.next.next is not None:
            curr = curr.next
        val = curr.next.val
        curr.next = None
        self._size -= 1
        return val

    def delete_at(self, index: int) -> int:
        if self.head is None:
            raise IndexError("list is empty")
        if index < 0 or index >= self._size:
            raise IndexError("index out of range")
        if index == 0:
            return self.delete_at_head()
        curr = self.head
        for _ in range(index - 1):
            curr = curr.next
        val = curr.next.val
        curr.next = curr.next.next
        self._size -= 1
        return val

    def search(self, val: int) -> int:
        curr = self.head
        i = 0
        while curr is not None:
            if curr.val == val:
                return i
            curr = curr.next
            i += 1
        return -1

    def reverse(self) -> None:
        prev = None
        curr = self.head
        while curr is not None:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        self.head = prev

    def __len__(self) -> int:
        return self._size

    def to_list(self) -> list[int]:
        result = []
        curr = self.head
        while curr is not None:
            result.append(curr.val)
            curr = curr.next
        return result


# ── Doubly Linked List ────────────────────────────────────────

class DoublyNode:
    def __init__(self, val: int = 0, prev: DoublyNode | None = None, next: DoublyNode | None = None):
        self.val = val
        self.prev = prev
        self.next = next


class DoublyLinkedList:
    def __init__(self):
        self.head: DoublyNode | None = None
        self.tail: DoublyNode | None = None
        self._size = 0

    def insert_at_head(self, val: int) -> None:
        node = DoublyNode(val, next=self.head)
        if self.head is not None:
            self.head.prev = node
        else:
            self.tail = node
        self.head = node
        self._size += 1

    def insert_at_tail(self, val: int) -> None:
        node = DoublyNode(val, prev=self.tail)
        if self.tail is not None:
            self.tail.next = node
        else:
            self.head = node
        self.tail = node
        self._size += 1

    def delete_at_head(self) -> int:
        if self.head is None:
            raise IndexError("list is empty")
        val = self.head.val
        self.head = self.head.next
        if self.head is not None:
            self.head.prev = None
        else:
            self.tail = None
        self._size -= 1
        return val

    def delete_at_tail(self) -> int:
        if self.tail is None:
            raise IndexError("list is empty")
        val = self.tail.val
        self.tail = self.tail.prev
        if self.tail is not None:
            self.tail.next = None
        else:
            self.head = None
        self._size -= 1
        return val

    def search(self, val: int) -> int:
        curr = self.head
        i = 0
        while curr is not None:
            if curr.val == val:
                return i
            curr = curr.next
            i += 1
        return -1

    def reverse(self) -> None:
        curr = self.head
        while curr is not None:
            curr.prev, curr.next = curr.next, curr.prev
            curr = curr.prev
        self.head, self.tail = self.tail, self.head

    def __len__(self) -> int:
        return self._size

    def to_list(self) -> list[int]:
        result = []
        curr = self.head
        while curr is not None:
            result.append(curr.val)
            curr = curr.next
        return result
